### ISSUE TITLE
CNAME records not resolved in link blocks when DNAME is present

### ISSUE THREAD

**Issue by esum** [inspircd/inspircd#1870]
**Description**

While resolving a domain name in a link block, InspIRCd seems to fail to resolve the name when a CNAME when a DNAME is also answered by the DNS server (DNAME should be transparent as a CNAME is also returned in the DNS answer).

**Steps to reproduce the issue:**

Create a link block with a name pointing to a DNAME record.

**Describe the results you received:**

InspIRCd is unable to fully resolve the domain name
SNOMASK +l sends the following message `Unable to resolve hostname - Malformed answer`

**Describe the results you expected:**

InspIRCd should be able to resolve the domain name behind the CNAME regardless of the DNAME presence in the answer.
It seems InspIRCd 2 was able to resolve the CNAME even with the DNAME present.

**Output of `./bin/inspircd --version`:**

InspIRCd-3.9.0

**Comment by mpaulon** [comment]
Looks like the [function](https://github.com/inspircd/inspircd/blob/ef23e4c07320fbbbf9d142878aae05a3e87eff0f/src/dns.cpp#L488
) CNAMEs resolution was dropped when moving from dns.cpp to core_dns.cpp

**Comment by Adam-** [comment]
I don't think this has anything to do with the core_dns move, if you look at insp20 it always did an AAAA <https://github.com/inspircd/inspircd/blob/insp20/src/modules/m_spanningtree/main.cpp#L275> followed by an A <https://github.com/inspircd/inspircd/blob/insp20/src/modules/m_spanningtree/resolvers.cpp#L75> when that fails. Which is what 3 currently does too. *Usually* the resolver will resolve the CNAME and include the A/AAAA records additionally in the response like so:

```
$ dig AAAA irc.inspircd.org

; <<>> DiG 9.16.1-Ubuntu <<>> AAAA irc.inspircd.org
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 58321
;; flags: qr rd ad; QUERY: 1, ANSWER: 3, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;irc.inspircd.org.              IN      AAAA

;; ANSWER SECTION:
irc.inspircd.org.       0       IN      CNAME   irc.chatspike.net.
irc.chatspike.net.      0       IN      AAAA    2a03:b0c0:1:d0::186:f001
irc.chatspike.net.      0       IN      AAAA    2001:41d0:602:2ed::f009
```

Though I'm not sure if that is required or not by any spec. The resolver in m_spanningtree then works just by looking for answers of the same type as the query:
```cpp
void SecurityIPResolver::OnLookupComplete(const DNS::Query *r)
{
        for (std::shared_ptr<Link> L : Utils->LinkBlocks)
        {
                if (L->IPAddr == host)
                {
                        for (const auto& ans_record : r->answers)
                        {
                                if (ans_record.type == this->question.type)
                                        Utils->ValidIPs.push_back(ans_record.rdata);
                        }
                        break;
                }
        }
}
```

**Comment by esum** [comment]
In fact it only happends when a DNAME is also present in the answer section, I edited the title and issue to make that clear.

**Comment by Adam-** [comment]
fixed in 030eaf9e574d0045265d11638e8cced4861423a7
